### PR TITLE
Add `key_storage`, `isolation_level` and `lock_manager` configuration to cache policies

### PR DESCRIPTION
**Pull request by desertaxle**
This PR adds the ability to specify a storage location for cache entries via a `CachePolicy`. The isolation level for the cache and locking mechanism are also configurable via a `CachePolicy`.

## Example
Cache task results based on the input and source code a task. Store those cache entries at `~/.prefect/cache`. Allow only one actor to access a cache entry at a time and store lock files to enforce that at `~.prefect/locks`.
```python
from prefect import task
from prefect.cache_policies import INPUTS, TASK_SOURCE
from prefect.transactions import IsolationLevel


@task(
    cache_policy=(INPUTS + TASK_SOURCE).configure(
        key_storage="~/.prefect/cache",
        lock_manager="~/.prefect/locks",
        isolation_level=IsolationLevel.SERIALIZABLE,
    )
)
def my_task(x):
    return x


if __name__ == "__main__":
    print(my_task(10))
```

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] If this pull request adds new functionality, it includes unit tests that cover the changes
- [ ] If this pull request removes docs files, it includes redirect settings in `mint.json`.
- [ ] If this pull request adds functions or classes, it includes helpful docstrings.

### CODE REVIEW

File: src/prefect/cache_policies.py [hunk 1]
```diff
@@ -21,6 +38,37 @@ def from_cache_key_fn(
...
+                the current storage will be used.
+            locks: The lock manager or storage path to use for the cache policy. If not provided,
+                the current lock manager will be used.
+            isolation_level: The isolation level to use for the cache policy. If not provided,
+                the current isolation level will be used.
+
+        Returns:
+            A new cache policy with the given storage, locks, and isolation level.
+        """
+        new = deepcopy(self)
```
Comment (cicdw) [inline src/prefect/cache_policies.py:63]: jw why is this necessary

File: src/prefect/transactions.py [hunk 2]
```diff
@@ -195,7 +195,9 @@ def __enter__(self):
             and not self.store.supports_isolation_level(self.isolation_level)
         ):
             raise ValueError(
-                f"Isolation level {self.isolation_level.name} is not supported by provided result store."
+                f"Isolation level {self.isolation_level.name} is not supported by provided "
```
Comment (cicdw) [inline src/prefect/transactions.py:198]: note for future: I think we should refactor many of these errors into special `prefect.exceptions.ConfigurationError` exceptions for clarity and consistency

File: src/prefect/cache_policies.py [hunk 3]
```diff
@@ -1,17 +1,34 @@
...
     Base class for all cache policies.
     """
 
+    storage: Union["WritableFileSystem", str, Path, None] = None
+    isolation_level: Union[
+        Literal["READ_COMMITTED", "SERIALIZABLE"],
+        "IsolationLevel",
+        None,
+    ] = None
+    locks: Union[str, Path, "LockManager", None] = None
```
Comment (cicdw) [inline src/prefect/cache_policies.py:30]: this kwarg doesn't 100% read right to me, as `lock` is such a strong noun already. `lock_manager` might be better despite being longer? any other options?

**Review by cicdw — COMMENTED** [review]
One note on `locks` as a kwarg name, otherwise this is great

File: src/prefect/cache_policies.py (hunk 1 above)
Comment (desertaxle) [reply src/prefect/cache_policies.py:63]: I was following the pattern of returning an new object from `CachePolicy` operations, but this method should probably return `self`. I can make that change.

File: src/prefect/cache_policies.py (hunk 3 above)
Comment (desertaxle) [reply src/prefect/cache_policies.py:30]: I called it `locks` because I was also allowing a file path as syntactic sugar for a `FileSystemLockManager`. I can try and think of an alternative name. Otherwise, I can remove that sugar, rename the kwarg to `lock_manager`, and accept just a `LockManager`.

File: src/prefect/cache_policies.py (hunk 1 above)
Comment (desertaxle) [reply src/prefect/cache_policies.py]: Updated to return update in place and return `self`. It works well but can lead to unexpected results if you call configure on one of the singleton cache policies like `INPUTS` or `DEFAULT`.

File: src/prefect/cache_policies.py (hunk 3 above)
Comment (desertaxle) [reply src/prefect/cache_policies.py]: I updated this to `lock_manager` and removed the option to pass a path. Let me know what you think.

File: src/prefect/cache_policies.py (hunk 1 above)
Comment (cicdw) [reply src/prefect/cache_policies.py]: Ahhhhh great catch -- since we do pre-instantiate these classes I think we should revert to your original instinct of deep copying self.

File: src/prefect/cache_policies.py [hunk 4]
```diff
@@ -1,17 +1,33 @@
...
     Base class for all cache policies.
     """
 
+    storage: Union["WritableFileSystem", str, Path, None] = None
+    isolation_level: Union[
+        Literal["READ_COMMITTED", "SERIALIZABLE"],
+        "IsolationLevel",
+        None,
+    ] = None
+    lock_manager: Optional["LockManager"] = None
```
Comment (cicdw) [inline src/prefect/cache_policies.py:29]: yea it's longer but I do like this more - very clear

File: src/prefect/cache_policies.py [hunk 5]
```diff
@@ -1,17 +1,33 @@
...
+    from prefect.transactions import IsolationLevel
+
 
 @dataclass
 class CachePolicy:
     """
     Base class for all cache policies.
     """
 
+    storage: Union["WritableFileSystem", str, Path, None] = None
```
Comment (cicdw) [inline src/prefect/cache_policies.py:23]: Only because now is the time to ask these questions - are there any other name options we have for this one, just to make crystal clear it's not the same thing as result storage?

File: src/prefect/cache_policies.py [hunk 6]
```diff
@@ -21,6 +37,36 @@ def from_cache_key_fn(
...
+        Returns:
+            A new cache policy with the given storage, locks, and isolation level.
+        """
+        if storage is not None:
+            self.storage = storage
+        if lock_manager is not None:
+            self.lock_manager = lock_manager
+        if isolation_level is not None:
+            self.isolation_level = isolation_level
+        return self
```
Comment (cicdw) [inline src/prefect/cache_policies.py:68]: commented elsewhere but your note about mutating our pre-instantiated policies was a good callout; I agree with your original implementation that we shouldn't mutate those and should instead start this off with a deepcopy

**Review by cicdw — CHANGES_REQUESTED** [review]
this looks awesome -- left one small comment about the `deepcopy` thing, and one question that isn't super important but I wanted to give it some airtime just in case. **cc:** @zzstoatzz in case you have thoughts about the kwarg name `storage`

File: src/prefect/cache_policies.py (hunk 5 above)
Comment (desertaxle) [reply src/prefect/cache_policies.py:23]: Another possibility could be `cache_storage`. It'd be a tad redundant, but it'd be explicit.

Comment (cicdw) [reply src/prefect/cache_policies.py:23]: A few other ideas just to explore our options:
- `key_storage`
- `metadata_storage` (mirrors the internal API)
- `key_files`
- `metadata_files`
- `cache_files`
- `cache_data`
- `metadata`

**Review by cicdw — APPROVED** [review]
